File: leet-code/238. Product of Array Except Self/238_product_of_array_except_self.py
def product_except_self(nums):
    answer = []

    p = 1

    for i in range(len(nums)):
        answer.append(p)
        p *= nums[i]

    p = 1

    for i in range(len(nums) - 1, -1, -1):
        answer[i] *= p
        p *= nums[i]

    return answer


print(product_except_self([1, 2, 3, 4]))
print(product_except_self([-1, 1, 0, -3, 3]))

# 1 1 2 6
# 24 12 4 1
# 24 12 8 6

File: leet-code/238. Product of Array Except Self/238_product_of_array_except_self.dart
class Solution {
  List<int> productExceptSelf(List<int> nums) {
    List<int> answer = [];

    int sums = 1;

    for (int i = 0; i < nums.length; i++) {
      answer.add(sums);
      sums *= nums[i];
    }

    sums = 1;

    for (int i = nums.length - 1; i > -1; i--) {
      answer[i] *= sums;
      sums *= nums[i];
    }

    return answer;
  }
}

void main() {
  Solution solution = Solution();
  print(solution.productExceptSelf([1, 2, 3, 4]));
}
